### ISSUE TITLE
How to Add Authorization with Bearer Token to Soap Request -> Add Soap Action

### ISSUE THREAD

**Issue by panmona** [pmorelli92/SoapHttpClient#7]
I want to use your SoapHttpClient for making Soap Request but I have difficulties in setting it up. Biggest hurdle is adding a Bearer Token to the SOAP Request.

In the SOAP UI I can add a Bearer token like this:
![image](https://user-images.githubusercontent.com/16801528/35324769-07717760-00f2-11e8-8930-9b013017e5fa.png)

How can I do the same thing trough code? (@RichardSlater @pmorelli92 @pmorelli-origin)

---
What I tried in steps:
1. Generating the Token trough the TokenClient of IdentityModel package.
2. Add it as a header like this:
```
var header = new XElement("Authorization", $"Bearer {token}");
```
3. Sending the request.
```
using (var soapClient = new SoapClient())
{
    var result =
        await soapClient.PostAsync(
            endpoint: endpoint,
            soapVersion: SoapVersion.Soap11,
            body: body,
            header: header);

            Log.Information(await result.Content.ReadAsStringAsync());
}
```

4. Inspecting the result
`Internal Server Error`
I think this is most probably because I didn't add the token in the right way.

**Comment by panmona** [comment]
I tried to remove the Token Authorization now from my Service and after removing the header from my call I still get the `Internal Server Error`. The call works perfectly in the SoapUI. 
What can I do to better debug the error?

**Comment by pmorelli92** [comment]
Hi @maracuja-juice, have you tried setting the header in the delegate of the HttpClient? 

```
var httpClient = new HttpClient();
httpClient.DefaultRequestHeaders.Authorization = ....
var soapClient = new SoapClient(() => httpClient)
```

**Comment by panmona** [comment]
Hi @pmorelli92 
Thanks for answering. No I haven't.

Currently I have no Authorization in my SOAP Service and I also get the `Internal Server Error`. 
What can I do to better debug this error as it isn't only the token?

**Comment by pmorelli92** [comment]
@maracuja-juice I would suggest installing Fiddler in order to see how is the call being made to the service.

**Comment by panmona** [comment]
@pmorelli92 
Okay. What I found out trough fiddler:
- The XML body is not the issue (the XML that my app sends works fine when sending the same XML with the Soap UI)
- The request headers seem to be the issue.

These are the request header that get sent from my app trough the SoapHttpClient:

![image](https://user-images.githubusercontent.com/16801528/35503890-bd9bf824-04e1-11e8-9a4d-7ddde2342e76.png)

And these are the request headers that get sent from the Soap UI:

![image](https://user-images.githubusercontent.com/16801528/35503950-e9719472-04e1-11e8-9bb6-68469d61e278.png)

The difference are:
- SoapUI uses Content-Length: 236 instead of 240
- SoapUI sends a User Agent
- SoapUI sends a Soap Action

I suspect that it's the SoapAction header that makes the difference. 

**How can I add this header to my request when using the PostAsync method?**

**Comment by pmorelli92** [comment]
@maracuja-juice are you sending the action parameter in the post async call?

**Comment by panmona** [comment]
@pmorelli92 Good point.

Do you mean like this?
```
var result = await soapClient.PostAsync(
                    endpoint: endpoint, 
                    soapVersion: SoapVersion.Soap11, 
                    body: body,
                    action: "IsAlive"); // Or with the full URL.
```
That doesn't add a SoapAction to the Request Headers but an ActionHeader.

**Comment by pmorelli92** [comment]
Then you can do something like:

```
var httpClient = new HttpClient();
httpClient.DefaultRequestHeaders.TryAddWithoutValidation("SOAPAction", "foobar");
var soapClient = new SoapClient(() => httpClient)
```

**Comment by panmona** [comment]
@pmorelli92 
Oh cool that works if I put the full URL like in the request of the SOAP Ui there.

I don't have much experience with SOAP but isn't that SOAPAction header something that should be added per default?

**Comment by pmorelli92** [comment]
I am going to expose an overload for adding the action in the client.PostAsync

**Comment by maracuja-juice** [comment]
Sounds good. In that overload do you plan that we need to add the full URL of the action?

**Comment by panmona** [comment]
Nice! 👍  When will you do that?

**Comment by pmorelli92** [comment]
@maracuja-juice sorry for the delay, this is because of #8. It should be working on the just released version 2.2 using the action parameter of the post signature.
